clang/lib/AST/ASTContext.cpp

```diff
@@ -3140,13 +3140,35 @@ const FunctionType *ASTContext::adjustFunctionType(const FunctionType *T,
   return cast<FunctionType>(Result.getTypePtr());
 }
 
+QualType ASTContext::adjustFunctionResultType(QualType FunctionType,
+                                              QualType ResultType) {
+  // Might be wrapped in a macro qualified type.
+  if (const auto *MQT = dyn_cast<MacroQualifiedType>(FunctionType)) {
+    return getMacroQualifiedType(
+        adjustFunctionResultType(MQT->getUnderlyingType(), ResultType),
+        MQT->getMacroIdentifier());
+  }
+
+  // Might have a calling-convention attribute.
+  if (const auto *AT = dyn_cast<AttributedType>(FunctionType)) {
+    return getAttributedType(
+        AT->getAttrKind(),
+        adjustFunctionResultType(AT->getModifiedType(), ResultType),
+        adjustFunctionResultType(AT->getEquivalentType(), ResultType));
+  }
+
+  // Anything else must be a function type. Rebuild it with the new return
+  // value.
```

Comment:
Is this assumption true? I added the two to my original PR because that's the two I have seen but no confidence.

Reply:
There is at least one other case that we might want to handle (`BTFTagAttributedType`), but I have so far been unsuccessful in trying to actually create a declaration that has this attribute; I also have no idea what that attribute is for.

Reply:
I think what would be required for this type of work is to have the general ability to traverse `Type`s and `TypeLoc`s. Implementation like this + the one I provided in `TreeTransform::TransformLambdaExpr` is error prone and wouldn't scale.

Reply:
I’ve been thinking about this too; something like that or some way to strip `TypeLoc`s that count as ‘sugar’; we’re doing similar things all over the place, but from what I can tell, what is or isn’t stripped is highly dependent on context.

Reply:
There’s `TypeLoc::getAsAdjusted()`; that might work for at least the `TransformLambdaExpr` case.

Reply:
Yup, that works in tree transform, so that we shouldn’t have to worry about that part anymore.

Reply:
Also, I just noticed, `RecursiveASTVisitor` *does* seem to support `Type`s and `TypeLoc`s? I’ll take another look at that.

Reply:
The thing is, I’m not sure how well an AST visitor would work for this specific use case (rebuilding a function type with a different result type).

Reply:
Also, I just noticed, this code seems to be modelled after `getFunctionTypeWithExceptionSpec()`, which is like two functions down in the same file and which does pretty much the same thing that this is doing here. It might make sense to merge the two.

Reply:
> Also, I just noticed, this code seems to be modelled after `getFunctionTypeWithExceptionSpec()`, which is like two functions down in the same file and which does pretty much the same thing that this is doing here. It might make sense to merge the two.

I was hacking just enough to make it work, apologies.

Reply:
Just refactored this and extracted the code that handles the sugar into a separate function (see the latest commit).